### PR TITLE
Examples: Add proper Table of Contents

### PR DESCRIPTION
**Pull request by AnnsAnns**
### Contribution description

Currently there is no description of the examples (Some not even having a README). 

This PR aims to add a basic table of contents README for the examples directory to improve the UX when searching for examples while also adding a few minor READMEs for undocumented examples.

It also tries to improve the general readability of the folder by categorizing examples into easier to digest categories within the README. This makes specific examples easier to find.

### Testing procedure

There are no changes within this PR that would affect the code.

### Issues/PRs references

This aims to help with some of the issues mentioned within #20388

### CODE REVIEW

File: examples/README.md [hunk 1]
```diff
@@ -0,0 +1,185 @@
+# Examples
+
+RIOT provides a wide range of examples to demonstrate the capabilities of the OS and its modules.
+
+The examples are organized in subdirectories of the `examples` directory. Each example directory contains a `README.md` file that provides information about the example and how to run it.
```
Comment (Teufelchen1) [inline examples/README.md:5]:
```suggestion
Each example contains a `README.md` that provides information and instructions how to run it.
```

Feels redundant?

File: examples/README.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+Here is a quick overview of the examples available in the RIOT repository:
```
Comment (Teufelchen1) [inline examples/README.md:7]:
```suggestion
Here is a quick overview of the examples available in the RIOT:
```

File: examples/README.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+  - A simple "Hello World" that shows the basic structure of a RIOT application.
+- [blinky](./blinky)
+  - The classic "Hello World" example for embedded systems: blinking an LED (or printing "Blink!" via stdio when none are available).
+- [leds_shell](./leds_shell)
+  - The application `leds_shell` is a basic example, which allows easy, interactive
+    control of internal board LEDs, and basic GPIO for externally connected simple
+    devices (for e.g. additional LEDs, relay, motors - via dedicated drivers, etc.)
+    via the shell.
+- [saul](./saul)
+  - This example demonstrates the usage of the SAUL (Sensor Abstraction Layer User) module.
```
Comment (Teufelchen1) [inline examples/README.md:25]: That's not SAUL ;-) https://doc.riot-os.org/group__drivers__saul.html

File: examples/README.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+#### RIOT Language Bindings
+
+##### Officially Supported/Targeted
+
+###### _Rust_
+
+- [rust-hello-world](./rust-hello-world)
+  - This example demonstrates how to write a simple RIOT application in Rust.
+- [rust-gcoap](./rust-gcoap)
+  - This example demonstrates how to write a simple RIOT application in Rust using the gcoap module.
```
Comment (Teufelchen1) [inline examples/README.md:42]:
```suggestion
  - This example demonstrates how to write a coap server application in Rust using the RIOTs gcoap module.
```
Maybe that's better?

File: examples/README.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+- [rust-async](./rust-async)
+  - This example demonstrates how to write a simple RIOT application in Rust using the async/await syntax.
```
Comment (Teufelchen1) [inline examples/README.md:44]:
```suggestion
  - This example demonstrates how to use Rusts async/await syntax in a RIOT application.
```

File: examples/pio_blink/README.md
Comment (Teufelchen1) [inline examples/pio_blink/README.md:1]: Misses instructions on how to run. (Mandated by yourself in the examples/Readme.md)

File: examples/telnet_server/README.md
Comment (Teufelchen1) [inline examples/telnet_server/README.md:1]: Misses instructions on how to run. (Mandated by yourself in the examples/Readme.md)

**Review by Teufelchen1 — CHANGES_REQUESTED** [review]
Didn't have time to do a throughout review but already found little issues.

File: examples/README.md (hunk 3 above)
Comment (AnnsAnns) [reply examples/README.md:25]: SALU 😄 Good catch

File: examples/pio_blink/README.md
Comment (AnnsAnns) [reply examples/pio_blink/README.md:1]: Will add

File: examples/telnet_server/README.md
Comment (AnnsAnns) [reply examples/telnet_server/README.md:1]: will add

**Review by AnnsAnns — COMMENTED** [review]
Will improve things mentioned, thank you for the review :)

**Comment by leandrolanzieri** [comment]
@AnnsAnns do you think it would be possible to include a basic static test that goes over the directory list in `examples` and checks that at least they're mentioned in the README.md? I'm afraid this index can age if not enforced somehow...

File: examples/timer_periodic_wakeup/README:md [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+## Overview
+
+The application sets up a timer that periodically wakes up the system and performs a specific task. This is useful for applications that need to perform regular actions, such as sensor readings or data transmissions.
+
+## Requirements
+
+- RIOT OS
+- A supported hardware platform
+
+## Setup
```
Comment (mguetschow) [inline examples/timer_periodic_wakeup/README:md:14]: This thorough step-by-step guide could in the future serve as a template for (automatically generated?) documentation about the examples at doc.riot-os.org. Feels a bit wordy compared to the other READMEs in the repo, but since they are all differing in detail anyway, I don't mind it too much.

**Review by mguetschow — CHANGES_REQUESTED** [review]
Thanks for your work on this! Just had a brief look and left some comments below.

Could we maybe have a two-column table-style of the list? At least, I would argue for moving the explanation from the single subitem to the item itself, potentially after a colon (`:`).

**Review by krzysztof-cabaj — COMMENTED** [review]
I think that there is a little typo for `timer_periodic_wakeup' and this file should be named README.md not README:md (period instead of colon).

File: examples/README.md [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+- [psa_crypto](./psa_crypto)
+
+  - Basic functions of the PSA Crypto API- [nimble_heart_rate_sensor](./nimble_heart_rate_sensor)
+  - This example demonstrates how to implement asynchronous data transfer using GATT
+    notifications by implementing a mock-up BLE heart rate sensor.
+
+- [pio_blink](./pio_blink)
+  - How to use the PIO peripheral on the RaspberryPi Pico to blink an LED.
+- [arduino_hello-world](./arduino_hello-world)
+  - This application demonstrates the usage of Arduino sketches in RIOT.
```
[the remark this answers is not in the document]
Comment (AnnsAnns) [reply examples/README.md:181]: Fair point

File: examples/README.md [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+- [lua_REPL](./lua_REPL)
+  - This example demonstrates how to use the Lua Read-Eval-Print Loop (REPL) in RIOT.
+- [micropython](./micropython)
+  - How to use the MicroPython port for RIOT.
+- [wasm](./wasm)
+  - How to use WebAssembly in RIOT.
+
+### Networking & BLE Examples
+
+#### Constraint Application Protocol (CoAP)
```
[the remark this answers is not in the document]
Comment (AnnsAnns) [reply examples/README.md:66]: Thanks :)

File: examples/README.md [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+RIOT provides a wide range of examples to demonstrate the capabilities of the OS and its modules.
+
+Each example contains a `README.md` that provides information and instructions how to run it.
+
+Here is a quick overview of the examples available in the RIOT:
+
+### The Essentials
+
+- [default](./default)
```
[the remark this answers is not in the document]
Comment (AnnsAnns) [reply examples/README.md:11]: Done

File: examples/README.md [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+- [lua_basic](./lua_basic)
+  - How to write IoT applications using Lua.
+- [lua_REPL](./lua_REPL)
+  - This example demonstrates how to use the Lua Read-Eval-Print Loop (REPL) in RIOT.
+- [micropython](./micropython)
+  - How to use the MicroPython port for RIOT.
+- [wasm](./wasm)
+  - How to use WebAssembly in RIOT.
+
+### Networking & BLE Examples
```
[the remark this answers is not in the document]
Comment (AnnsAnns) [reply examples/README.md]: Done

File: examples/README.md [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+
+#### Bluetooth Low Energy (BLE)
+
+- [nimble_scanner](./nimble_scanner)
+  - This example showcases the usage of the `NimBLE` BLE stack as a scanner.
+- [skald_eddystone](./skald_eddystone)
+  - This example demonstrates the usage of `Skald` for creating an Google `Eddystone` beacon.
+- [skald_ibeacon](./skald_ibeacon)
+  - This example demonstrates the usage of `Skald` for creating an Apple `iBeacon`.
+- [nimble_gatt](./nimble_gatt)
```
[the remark this answers is not in the document]
Comment (AnnsAnns) [reply examples/README.md]: done

File: examples/README.md [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+  - This example demonstrates how to monitor energy levels on all available wireless channels
+- [sniffer](./sniffer)
+  - This application is built to run together with the script `./tools/sniffer.py`
+    as a sniffer for (wireless) data traffic.
+- [dtls-sock](./dtls-sock)
+  - This example shows how to use DTLS sock `sock_dtls_t`
+- [dtls-echo](./dtls-echo)
+  - This example shows how to use TinyDTLS with sock_udp.
+- [dtls-wolfssl](./dtls-wolfssl)
+  - This example demonstrates the usage of the DTLS module with the wolfSSL library.
```
[the remark this answers is not in the document]
Comment (AnnsAnns) [reply examples/README.md]: done

File: examples/README.md [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+  - This example shows how to integrate SUIT-compliant firmware updates into a
+    RIOT application.
+- [thread_duel](./thread_duel)
+  - This is a thread duel application to show RIOTs abilities to run multiple-threads
+    concurrently, even if they are neither cooperative nor dividable into different scheduler priorities, by using the optional round-robin scheduler module.
+- [posix_select](./posix_select)
+  - This example is a showcase for RIOT's POSIX select support
+- [psa_crypto](./psa_crypto)
+
+  - Basic functions of the PSA Crypto API- [nimble_heart_rate_sensor](./nimble_heart_rate_sensor)
```
[the remark this answers is not in the document]
Comment (AnnsAnns) [reply examples/README.md]: done

**Comment by AnnsAnns** [comment]
> @AnnsAnns do you think it would be possible to include a basic static test that goes over the directory list in `examples` and checks that at least they're mentioned in the README.md? I'm afraid this index can age if not enforced somehow...

Added (Even found a few missing entries thanks to it :sweat_smile:)

**Comment by AnnsAnns** [comment]
I hope I covered all mentioned issues, let me know if there is anything else :)

**Review by Teufelchen1 — APPROVED** [review]

**Comment by Teufelchen1** [comment]
Nice! I am happy with it. There's one static test that you should fix. A nit pick could be to combine the two ci tests into one, reducing the burden on the servers but I believe that is not significant here.

**Review by mguetschow — APPROVED** [review]
Looks good, thank you!

**Comment by mguetschow** [comment]
Would you mind squashing the commits together?

**Comment by AnnsAnns** [comment]
> Did you maybe accidentally drop some commit?
> 
> ```
> ./dist/tools/examples_check/check_has_readme.sh x
>   Command output:
>   
>   	The following directories are missing a README:
>   	- pio_blink
>   
> ./dist/tools/examples_check/check_in_readme.sh x
>   Command output:
>   
>   	The following directories are missing in the README.md file:
>   	- sock_tcp_echo
> ```

Fixed

**Comment by mguetschow** [comment]
@AnnsAnns we've merged a new example just today with #20024, and your CI check rightfully noticed that. Would you mind adding it to the table of content?

**Comment by AnnsAnns** [comment]
> @AnnsAnns we've merged a new example just today with #20024, and your CI check rightfully noticed that. Would you mind adding it to the table of content?

Done

**Comment by Teufelchen1** [comment]
Congrats! It's really awesome!
